### PR TITLE
refactor out existing AI functions cleanly into "Strikes" and "Targets"

### PR DESCRIPTION
**Pull request by amannm**
- improve precision, performance of positioning and sampling computation involved in strikes
- properly release references to destroyed target entities to allow GC

### Brief description of the changes:

**[Interactive demo](http://jsfiddle.net/mcyt3hfx/)
Click to switch between strike models (old vs new)**

here is the current strike behavior that this is replacing:
Screen Shot 2022-09-11 at 11 13 58 AM

if the target selected is the center of the green circle, the current code is picking a random one of those five angles, then picking a random position on the black strips to move biters to before attacking the 32 radius green area

so with these changes you'll get a wider, path-dependent, continuous strike point distribution as seen below. 
Screen Shot 2022-09-11 at 11 14 22 AM

the "source" position (where the biter unit spawns) influences the strike start position. I'm doing extra geometry calculations to avoid picking positions requiring travel through the "clearance zone" of the target they will be ultimately attacking. 

sampling test fixture:

```lua
local boundary_offset = 20
local source = {x = -1024, y = 200}
local target = {x = 0, y = 200}
local samples = 2000
local res = ""
for _ = 1, samples, 1 do
    local strike_position = select_strike_position(source, target, boundary_offset)
    res = res .. string.format("[%s,%s],", math.floor(strike_position.x), math.floor(strike_position.y))
end
print(res)
```

### Tested Changes:
- [x] I've tested the changes locally or with people.
- [ ] I've not tested the changes.

### CODE REVIEW

**Review by TheBigZet — COMMENTED** [review]
Consider assigning frequently used functions (math.atan2, sqrt, abs) to local variables

**Comment by amannm** [comment]
> Consider assigning frequently used functions (math.atan2, sqrt, abs) to local variables

done

**Comment by amannm** [comment]
image
here's the distribution you see using the following test fixture:

```
local boundary_offset = 20
local source = {x = -1024, y = 200}
local target = {x = 0, y = 200}
local samples = 2000
local res = ""
for _ = 1, samples, 1 do
    local strike_position = select_strike_position(source, target, boundary_offset)
    res = res .. string.format("[%s,%s],", math.floor(strike_position.x), math.floor(strike_position.y))
end
print(res)

```
where source was varied with `{x = -1024, y = 200}` ` {x = 0, y = 1024}` ` {x = 1024, y = 200}`

**Comment by developer-8** [comment]
This was not tested for sure. Terrain generation very broken. You can move to draft if PR not finished

**Comment by amannm** [comment]
> This was not tested for sure. Terrain generation very broken. You can move to draft if PR not finished

sorry about that, had tested PR against main/master... develop introduced some new additional `Functions` use and the removal of the import messed that up. fixed now

**Comment by amannm** [comment]
here is the current strike behavior that this is replacing:
image
if the target selected is the center of the green circle, the current code is picking a random one of those three angles, then picking a random position on the grey portion to move biters to before attacking the 32 radius green area

**Comment by Ragnarok77-factorio** [comment]
This PR generates this spam in logs (ignore line number of screenshot) : 
image

Which is there : 
image

How to reproduce issue : 
Demote yourself, untrust yourself, place a ghost bp and welcome to this error where it will attempt to track an entity ghost that was basically deleted by code before.

==
I think its' just a missing an "if valid" condition in start_tracking code

File: maps/biter_battles_v2/ai_targets.lua [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+            local random_index = math_random(1, population_size)
+            local individual = population_list[random_index]
+            table_insert(sample, individual)
+        end
+    end
+    return sample
+end
+
+function Public.start_tracking(entity)
+    if target_entity_type[entity.type] and entity.unit_number then
```
Comment (Ragnarok77-factorio) [inline maps/biter_battles_v2/ai_targets.lua:57]: Missing an if valid check at the beginning (cf comment for why it's needed)

**Review by Ragnarok77-factorio — CHANGES_REQUESTED** [review]

**Comment by Ragnarok77-factorio** [comment]
Since sadly I think he's gone :( I made another PR so I can fix this issue : https://github.com/Factorio-Biter-Battles/Factorio-Biter-Battles/pull/340
(if he comes back I will cancel the new PR)

**Comment by TheBigZet** [comment]
@Ragnarok77-factorio

File: maps/biter_battles_v2/ai_targets.lua (hunk 1 above)
Comment (Ragnarok77-factorio) [reply maps/biter_battles_v2/ai_targets.lua:59]: done by TBZ, closed.

**Comment by Ragnarok77-factorio** [comment]
Thanks TBZ !  closed PR340

**Comment by amannm** [comment]
oh ok thanks for coming back to this, thanks for the fix, let me know if there's anything else that can be improved

**Review by Ragnarok77-factorio — APPROVED** [review]
Got a hard time to review code exactly line per line and fully understand it every small bit, but I  didn't find anything shocking at least... Feel free if someone else can deeply review it.

It was also tested on multiple games and didn't notice any issue anymore after fix, so ok for me.

**Review by TheBigZet — APPROVED** [review]
No issues reported during a month long testing

**Pull request merged**